File: monte_carlo_pi.py
import random
import time
import math

def estimate_pi(num_points):
    """
    Estimate Pi using Monte Carlo simulation.
    """
    print(f"Starting Monte Carlo Pi estimation with {num_points:,} points...")
    
    inside_circle = 0
    
    for i in range(num_points):
        # Generate random point in [0,1] x [0,1]
        x = random.random()
        y = random.random()
        
        # Check if point is inside quarter circle
        if x**2 + y**2 <= 1.0:
            inside_circle += 1
        
        # Progress updates every 50 million points
        if i % 50000000 == 0 and i > 0:
            progress = (i / num_points) * 100
            print(f"Progress: {i:,} points processed ({progress:.1f}%)")
    
    # Calculate Pi estimate
    pi_estimate = 4.0 * inside_circle / num_points
    
    return pi_estimate, inside_circle

if __name__ == "__main__":
    # Set random seed for reproducibility
    random.seed(42)
    
    # Number of random points
    num_points = 500000000
    
    # Run simulation
    start_time = time.time()
    pi_estimate, inside_circle = estimate_pi(num_points)
    elapsed_time = time.time() - start_time
    
    # Print results
    print("\n=== Results ===")
    print(f"Estimated Pi: {pi_estimate:.8f}")
    print(f"Actual Pi:    {math.pi:.8f}")
    print(f"Error:        {abs(pi_estimate - math.pi):.8f}")
    print(f"Used {num_points:,} points")
    print(f"Points inside circle: {inside_circle:,}")
    print(f"Elapsed time: {elapsed_time:.2f} seconds")

File: monte_carlo_pi.R
# Monte Carlo Pi Estimation in R
# This script estimates Pi using random sampling

estimate_pi <- function(num_points) {
  cat("Starting Monte Carlo Pi estimation with", format(num_points, big.mark=","), "points...\n")
  
  inside_circle <- 0
  
  for (i in 1:num_points) {
    # Generate random point in [0,1] x [0,1]
    x <- runif(1, 0, 1)
    y <- runif(1, 0, 1)
    
    # Check if point is inside quarter circle
    if (x^2 + y^2 <= 1) {
      inside_circle <- inside_circle + 1
    }
    
    # Progress updates every 50 million points
    if (i %% 50000000 == 0 && i > 0) {
      progress <- (i / num_points) * 100
      cat(sprintf("Progress: %s points processed (%.1f%%)\n", 
                  format(i, big.mark=","), progress))
    }
  }
  
  # Calculate Pi estimate
  pi_estimate <- 4 * inside_circle / num_points
  
  return(list(
    pi_estimate = pi_estimate,
    inside_circle = inside_circle,
    num_points = num_points
  ))
}

# Main execution
set.seed(42)  # For reproducibility
num_points <- 500000000

# Record start time
start_time <- Sys.time()

# Run simulation
result <- estimate_pi(num_points)

# Record end time
end_time <- Sys.time()
elapsed_time <- as.numeric(difftime(end_time, start_time, units = "secs"))

# Print results
cat("\n=== Results ===\n")
cat(sprintf("Estimated Pi: %.8f\n", result$pi_estimate))
cat(sprintf("Actual Pi:    %.8f\n", pi))
cat(sprintf("Error:        %.8f\n", abs(result$pi_estimate - pi)))
cat(sprintf("Used %s points\n", format(result$num_points, big.mark=",")))
cat(sprintf("Points inside circle: %s\n", format(result$inside_circle, big.mark=",")))
cat(sprintf("Elapsed time: %.2f seconds\n", elapsed_time))
